### ISSUE TITLE
What is the pose_retriever?

### ISSUE THREAD

**Issue by victkk** [CUT3R/CUT3R#13]
It seems to be used for initializing the pose token(denoted as z in the paper) prepended to the img token. The paper briefly mentioned this as "learnable pose token". Could you explain how exactly does the "learnable" work or pin a relevant paper?  Thank you.

**Comment by Kitsunetic** [comment]
I'm also exploring the code.

I believe you're referring to this part of the code:
https://github.com/CUT3R/CUT3R/blob/3cf9d8ee48f602e3937c6b26fb91465e707ca1cb/src/dust3r/model.py#L815-L854

It appears that there are two pose-related learnable parameters:
- `self.pose_retriever.mem`
- `self.pose_token`

`self.pose_token` serves as the pose feature for the first view.
`self.pose_retriever.mem` is is then iteratively updated using a memory module `LocalMemory`, which operates similarly to a RIN-style read-write network ([RIN Paper](https://arxiv.org/abs/2212.11972)):
https://github.com/CUT3R/CUT3R/blob/3cf9d8ee48f602e3937c6b26fb91465e707ca1cb/src/dust3r/model.py#L140-L222

Let me know if there's anything specific that remains unclear!

**Comment by victkk** [comment]
> I'm also exploring the code.
> 
> I believe you're referring to this part of the code:
> 
> [CUT3R/src/dust3r/model.py](https://github.com/CUT3R/CUT3R/blob/3cf9d8ee48f602e3937c6b26fb91465e707ca1cb/src/dust3r/model.py#L815-L854)
> 
> Lines 815 to 854 in [3cf9d8e](/CUT3R/CUT3R/commit/3cf9d8ee48f602e3937c6b26fb91465e707ca1cb)
> 
>  def _forward_impl(self, views, ret_state=False): 
>      shape, feat_ls, pos = self._encode_views(views) 
>      feat = feat_ls[-1] 
>      state_feat, state_pos = self._init_state(feat[0], pos[0]) 
>      mem = self.pose_retriever.mem.expand(feat[0].shape[0], -1, -1) 
>      init_state_feat = state_feat.clone() 
>      init_mem = mem.clone() 
>      all_state_args = [(state_feat, state_pos, init_state_feat, mem, init_mem)] 
>      ress = [] 
>      for i in range(len(views)): 
>          feat_i = feat[i] 
>          pos_i = pos[i] 
>          if self.pose_head_flag: 
>              global_img_feat_i = self._get_img_level_feat(feat_i) 
>              if i == 0: 
>                  pose_feat_i = self.pose_token.expand(feat_i.shape[0], -1, -1) 
>              else: 
>                  pose_feat_i = self.pose_retriever.inquire(global_img_feat_i, mem) 
>              pose_pos_i = -torch.ones( 
>                  feat_i.shape[0], 1, 2, device=feat_i.device, dtype=pos_i.dtype 
>              ) 
>          else: 
>              pose_feat_i = None 
>              pose_pos_i = None 
>          new_state_feat, dec = self._recurrent_rollout( 
>              state_feat, 
>              state_pos, 
>              feat_i, 
>              pos_i, 
>              pose_feat_i, 
>              pose_pos_i, 
>              init_state_feat, 
>              img_mask=views[i]["img_mask"], 
>              reset_mask=views[i]["reset"], 
>              update=views[i].get("update", None), 
>          ) 
>          out_pose_feat_i = dec[-1][:, 0:1] 
>          new_mem = self.pose_retriever.update_mem( 
>              mem, global_img_feat_i, out_pose_feat_i 
>          ) 
> It appears that there are two pose-related learnable parameters:
> 
> * `self.pose_retriever.mem`
> * `self.pose_token`
> 
> `self.pose_token` serves as the pose feature for the first view. `self.pose_retriever.mem` is is then iteratively updated using a memory module `LocalMemory`, which operates similarly to a RIN-style read-write network ([RIN Paper](https://arxiv.org/abs/2212.11972)):
> 
> [CUT3R/src/dust3r/model.py](https://github.com/CUT3R/CUT3R/blob/3cf9d8ee48f602e3937c6b26fb91465e707ca1cb/src/dust3r/model.py#L140-L222)
> 
> Lines 140 to 222 in [3cf9d8e](/CUT3R/CUT3R/commit/3cf9d8ee48f602e3937c6b26fb91465e707ca1cb)
> 
>  class LocalMemory(nn.Module): 
>      def __init__( 
>          self, 
>          size, 
>          k_dim, 
>          v_dim, 
>          num_heads, 
>          depth=2, 
>          mlp_ratio=4.0, 
>          qkv_bias=False, 
>          drop=0.0, 
>          attn_drop=0.0, 
>          drop_path=0.0, 
>          act_layer=nn.GELU, 
>          norm_layer=nn.LayerNorm, 
>          norm_mem=True, 
>          rope=None, 
>      ) -> None: 
>          super().__init__() 
>          self.v_dim = v_dim 
>          self.proj_q = nn.Linear(k_dim, v_dim) 
>          self.masked_token = nn.Parameter( 
>              torch.randn(1, 1, v_dim) * 0.2, requires_grad=True 
>          ) 
>          self.mem = nn.Parameter( 
>              torch.randn(1, size, 2 * v_dim) * 0.2, requires_grad=True 
>          ) 
>          self.write_blocks = nn.ModuleList( 
>              [ 
>                  DecoderBlock( 
>                      2 * v_dim, 
>                      num_heads, 
>                      mlp_ratio=mlp_ratio, 
>                      qkv_bias=qkv_bias, 
>                      norm_layer=norm_layer, 
>                      attn_drop=attn_drop, 
>                      drop=drop, 
>                      drop_path=drop_path, 
>                      act_layer=act_layer, 
>                      norm_mem=norm_mem, 
>                      rope=rope, 
>                  ) 
>                  for _ in range(depth) 
>              ] 
>          ) 
>          self.read_blocks = nn.ModuleList( 
>              [ 
>                  DecoderBlock( 
>                      2 * v_dim, 
>                      num_heads, 
>                      mlp_ratio=mlp_ratio, 
>                      qkv_bias=qkv_bias, 
>                      norm_layer=norm_layer, 
>                      attn_drop=attn_drop, 
>                      drop=drop, 
>                      drop_path=drop_path, 
>                      act_layer=act_layer, 
>                      norm_mem=norm_mem, 
>                      rope=rope, 
>                  ) 
>                  for _ in range(depth) 
>              ] 
>          ) 
>   
>      def update_mem(self, mem, feat_k, feat_v): 
>          """ 
>          mem_k: [B, size, C] 
>          mem_v: [B, size, C] 
>          feat_k: [B, 1, C] 
>          feat_v: [B, 1, C] 
>          """ 
>          feat_k = self.proj_q(feat_k)  # [B, 1, C] 
>          feat = torch.cat([feat_k, feat_v], dim=-1) 
>          for blk in self.write_blocks: 
>              mem, _ = blk(mem, feat, None, None) 
>          return mem 
>   
>      def inquire(self, query, mem): 
>          x = self.proj_q(query)  # [B, 1, C] 
>          x = torch.cat([x, self.masked_token.expand(x.shape[0], -1, -1)], dim=-1) 
>          for blk in self.read_blocks: 
>              x, _ = blk(x, mem, None, None) 
>          return x[..., -self.v_dim :] 
> Let me know if there's anything specific that remains unclear!

Thank you! It's clear enough!
